WordPress/gutenberg
# `Navigator`: add basic history support, enable advanced focus restoration

## Description


Following the conversation in #37063, this PR is an exploration into enabling focus restoration when navigating back

Changes:
- added basic location history stack (necessary for implementing focus restoration):
  - refactored the `Navigator` component, especially the shape of the context object: it now exposes `location` (the current location), `push` (used to navigate to a new location), and `pop` (used to navigate to the previous location)
  - refactored usages across gutenberg
- enhanced focus restoration:
  - prior to this PR, each `NavigatorScreen` would focus its first tabbable element when mounted
  - this PR adds logic to attempt to restore focus on a specific element when navigating back (with a fallback on the first tabbable element, as it already happens)
  - updated Storybook example to take advantage of that
  - note that, as of now, the element that will receive focus is targeted via `querySelector`, and it is the responsibility of the developer to make sure that each link can be correctly `selected` by providing a working selector.


TODO:
- edit Storybook example to add a standard `NavigatorLink`
- add/update `Navigator`'s unit tests
- update `Navigator` README and JSDocs (including the `@example`s)
- add CHANGELOG entry

## How has this been tested?




- [TODO] Unit tests
- Check that the Storybook example works as expected (both focus restoration and back/forward navigation)
- Check that the Global Styles Sidebar in the site editor works as expected (both focus restoration and back/forward navigation)
- Check that the Preferences Modal in the post editor (on mobile screens) works as expected (both focus restoration and back/forward navigation)

## Screenshots 

### Storybook

https://user-images.githubusercontent.com/1083581/145439161-fa003973-0c73-4f5d-95bd-8f8160f09ba8.mp4

### Preferences Modal

https://user-images.githubusercontent.com/1083581/145445546-c91306cf-be68-436c-b0e8-56243b5e481f.mp4

### Global Styles Sidebar

https://user-images.githubusercontent.com/1083581/145446524-aefcdd02-1294-4627-a43d-661bdf582bdb.mp4


## Types of changes





New feature (technically not breaking, since the component is experimental)

## Checklist:
